gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,514 @@
+    // The actual record constructor.
+    private final Constructor<T> constructor;
+    private final Object[] constuctorDefaultValues;
+
+    private RecordTypeAdapterImpl(RecordField[] recordFields, Constructor<T> constructor) {
+      this.recordFields = recordFields;
+      this.constructor = constructor;
+      for (int i = 0; i < recordFields.length; i++) {
+        RecordField recordField = recordFields[i];
+        for (String name : recordField.deSerializedNames) {
+          Integer prevIndex = recordFieldIndex.put(name, i);
+          if (prevIndex != null) {
+            throw new IllegalArgumentException(
+                "Both ["
+                    + recordFields[prevIndex]
+                    + "]"
+                    + " and ["
+                    + recordField
+                    + "] can be read from the same name ["
+                    + name
+                    + "]");
+          }
+        }
+      }
+      constuctorDefaultValues = new Object[recordFields.length];
+      for (int i = 0; i < recordFields.length; i++) {
+        constuctorDefaultValues[i] = recordFields[i].defaultValue;
+      }
+    }
+
+    @Override
+    public void write(JsonWriter out, T value) throws IOException {
+      if (value == null) {
+        out.nullValue();
+        return;
+      }
+      out.beginObject();
+      for (RecordField recordField : recordFields) {
+        recordField.appendNameAndValue(out, value);
+      }
+      out.endObject();
+    }
+
+    @Override
+    public T read(JsonReader in) throws IOException {
+      if (in.peek() == JsonToken.NULL) {
+        in.nextNull();
+        return null;
+      } else if (in.peek() != JsonToken.BEGIN_OBJECT) {
+        throw new DeserializeException("Expecting null or begin object at [" + in.getPath() + "]");
+      }
+      in.beginObject();
+      // This array will hold the value for each json field in the src object that we find,
+      // deserialized using a RecordField, that will also respect all Gson annotations. For fields
+      // that are not deserialized, the default null value will be used.
+      Object[] values = new Object[recordFields.length];
+      // Copy the default values, this will ensure that primitives are non-null.
+      System.arraycopy(constuctorDefaultValues, 0, values, 0, values.length);
+
+      while (in.peek() != JsonToken.END_OBJECT) {
+        String fieldName = in.nextName();
+        // Check if we know which field this name should deserialize to. If we do not find a
+        // matching index, then this field is one that is not present in our record, and ignored. We
+        // do not treat unknown object fields as an error, because the version of the class we have
+        // might differ from the source.
+        Integer fieldIndex = recordFieldIndex.get(fieldName);
+        if (fieldIndex != null) {
+          values[fieldIndex] = recordFields[fieldIndex].readValueForKey(in);
```

Comment:
Hi, I was also working on a record type adapter when I came across this PR. 
After playing around a bit with it, I think there is a missing null check here:
```suggestion
          Object value = recordFields[fieldIndex].readValueForKey(in);
          if(value != null){
              values[fieldIndex] = value
          }
```

Without this, the code overrides the default values for primitives with null.

Reply:
Possibly the `null` check should only be applied if the component type is a primitive to allow duplicate non-primitive members to overwrite each others values, to be consistent with `ReflectiveTypeAdapterFactory`'s behavior.

Reply:
> If you only want to continue working on the new pull request, it might be good at the end to have a look at the review comments here again nonetheless; maybe some of them also apply to the new pull request.

I think the new pull request is the way to go, as it is a bit more integrated with the normal reflection flow. I will still check these reviews to see if any apply. Thanks for the input.